File: routes/notes.py
from flask import Blueprint, flash, redirect, render_template, request, session, url_for
from sqlalchemy import or_

from auth_helpers import login_required
from extensions import db
from models import Note
from services import current_user_id, paginate_query, sanitize_plain, summarize_text, timestamp_now


notes_bp = Blueprint("notes", __name__)


@notes_bp.route("/notes_hub", methods=["GET", "POST"])
@login_required
def notes_hub():
    query = request.args.get("q", "").strip()
    show_archived = request.args.get("archived") == "1"
    page = request.args.get("page", 1, type=int)
    if request.method == "POST":
        title = sanitize_plain(request.form.get("title"))
        content = sanitize_plain(request.form.get("content"))
        subject = sanitize_plain(request.form.get("subject"))
        tags = sanitize_plain(request.form.get("tags"))
        is_public = request.form.get("visibility") == "public"
        if title and content:
            now = timestamp_now()
            db.session.add(
                Note(
                    user_id=session["user_id"],
                    title=title,
                    content=content,
                    subject=subject,
                    tags=tags,
                    created_at=now,
                    updated_at=now,
                    is_public=is_public,
                )
            )
            db.session.commit()
            flash("Note created successfully.", "success")
        else:
            flash("Title and content are required.", "error")
        return redirect(url_for("notes.notes_hub"))

    note_query = Note.query.filter_by(user_id=session["user_id"], is_archived=show_archived)
    if query:
        search = f"%{query}%"
        note_query = note_query.filter(
            or_(
                Note.title.ilike(search),
                Note.content.ilike(search),
                Note.subject.ilike(search),
                Note.tags.ilike(search),
            )
        )
    paginated_notes = paginate_query(
        note_query.order_by(Note.is_pinned.desc(), Note.updated_at.desc(), Note.created_at.desc()),
        page,
        12,
    )
    community_notes = Note.query.filter_by(is_public=True).order_by(Note.likes.desc()).limit(6).all()

    return render_template(
        "notes_hub.html",
        notes=paginated_notes["items"],
        pagination=paginated_notes,
        community_notes=community_notes,
        query=query,
        show_archived=show_archived,
    )


@notes_bp.route("/notes", methods=["GET", "POST"])
@login_required
def notes():
    if request.method == "POST":
        title = sanitize_plain(request.form.get("title"))
        content = sanitize_plain(request.form.get("content"))
        if title and content:
            db.session.add(
                Note(
                    user_id=current_user_id(),
                    title=title,
                    content=content,
                    created_at=timestamp_now(),
                    updated_at=timestamp_now(),
                    is_public=False,
                )
            )
            db.session.commit()
            flash("Note saved.", "success")
        else:
            flash("Title and content are required.", "error")
        return redirect(url_for("notes.notes"))

    page = request.args.get("page", 1, type=int)
    paginated_notes = paginate_query(
        Note.query.filter_by(user_id=current_user_id()).order_by(Note.created_at.desc()),
        page,
        12,
    )
    return render_template("notes.html", notes=paginated_notes["items"], pagination=paginated_notes, summary=None)


@notes_bp.route("/notes/<int:note_id>/edit", methods=["POST"])
@login_required
def edit_note(note_id):
    note = Note.query.filter_by(id=note_id, user_id=current_user_id()).first()
    if not note:
        flash("Note not found.", "error")
        return redirect(url_for("notes.notes_hub"))

    title = sanitize_plain(request.form.get("title"))
    content = sanitize_plain(request.form.get("content"))
    if not title or not content:
        flash("Title and content are required.", "error")
        return redirect(url_for("notes.notes_hub"))

    note.title = title
    note.content = content
    note.subject = sanitize_plain(request.form.get("subject"))
    note.tags = sanitize_plain(request.form.get("tags"))
    note.is_public = request.form.get("visibility") == "public"
    note.updated_at = timestamp_now()
    db.session.commit()
    flash("Note updated.", "success")
    return redirect(url_for("notes.notes_hub"))


@notes_bp.route("/notes/<int:note_id>/<action>", methods=["POST"])
@login_required
def update_note_state(note_id, action):
    note = Note.query.filter_by(id=note_id, user_id=current_user_id()).first()
    if not note:
        flash("Note not found.", "error")
        return redirect(url_for("notes.notes_hub"))

    if action == "favorite":
        note.is_favorite = not bool(note.is_favorite)
    elif action == "pin":
        note.is_pinned = not bool(note.is_pinned)
    elif action == "archive":
        note.is_archived = True
    elif action == "restore":
        note.is_archived = False
    elif action == "delete":
        db.session.delete(note)
        db.session.commit()
        flash("Note deleted.", "success")
        return redirect(url_for("notes.notes_hub"))
    else:
        flash("Unknown note action.", "error")
        return redirect(url_for("notes.notes_hub"))

    note.updated_at = timestamp_now()
    db.session.commit()
    flash("Note updated.", "success")
    return redirect(url_for("notes.notes_hub"))


@notes_bp.route("/notes/<int:note_id>/summary")
@login_required
def summarize_note(note_id):
    note = Note.query.filter_by(id=note_id, user_id=current_user_id()).first()
    if not note:
        flash("Note not found.", "error")
        return redirect(url_for("notes.notes"))

    paginated_notes = paginate_query(
        Note.query.filter_by(user_id=current_user_id()).order_by(Note.created_at.desc()),
        1,
        12,
    )
    return render_template(
        "notes.html",
        notes=paginated_notes["items"],
        pagination=paginated_notes,
        summary=summarize_text(note.content),
    )

File: lib/features/notes.ts
import { and, desc, eq, ilike, or } from "drizzle-orm";
import { revalidatePath } from "next/cache";
import { redirect } from "next/navigation";
import { db } from "@/db";
import { notes } from "@/db/schema";
import { currentUser } from "@/lib/auth/session";
import { verifyCsrfToken } from "@/lib/auth/csrf";
import { invalidateAcademicAutopilotCache, ownedAcademicSelection } from "@/lib/features/academic";
import { sanitizePlain } from "@/lib/text";

function redirectWith(message: string, type: "error" | "success" = "success"): never {
  redirect(`/notes?${type}=${encodeURIComponent(message)}`);
}

export async function createNoteAction(formData: FormData) {
  "use server";
  try {
    await verifyCsrfToken(formData);
  } catch {
    redirectWith("Security check failed. Please try again.", "error");
  }
  const user = await currentUser();
  if (!user) redirect("/login");
  const title = sanitizePlain(formData.get("title")).slice(0, 255);
  const content = sanitizePlain(formData.get("content"));
  const subject = sanitizePlain(formData.get("subject")).slice(0, 128) || null;
  const tags = sanitizePlain(formData.get("tags")).slice(0, 255) || null;
  const academic = await ownedAcademicSelection(user.id, formData.get("subject_id"), formData.get("topic_id"));
  if (!title || !content) redirectWith("Title and content are required.", "error");

  const now = new Date();
  await db.insert(notes).values({
    userId: user.id,
    subjectId: academic.subjectId,
    topicId: academic.topicId,
    title,
    content,
    subject: academic.subjectName ?? subject,
    tags,
    isPublic: formData.get("visibility") === "public",
    createdAt: now,
    updatedAt: now,
  });
  invalidateAcademicAutopilotCache(user.id);
  revalidatePath("/notes");
  revalidatePath("/autopilot");
  revalidatePath("/");
  redirectWith("Note created successfully.");
}

export async function updateNoteStateAction(formData: FormData) {
  "use server";
  try {
    await verifyCsrfToken(formData);
  } catch {
    redirectWith("Security check failed. Please try again.", "error");
  }
  const user = await currentUser();
  if (!user) redirect("/login");
  const noteId = Number(formData.get("note_id"));
  const action = sanitizePlain(formData.get("action"));
  const [note] = await db.select().from(notes).where(and(eq(notes.id, noteId), eq(notes.userId, user.id))).limit(1);
  if (!note) redirectWith("Note not found.", "error");

  if (action === "delete") {
    await db.delete(notes).where(eq(notes.id, note.id));
    invalidateAcademicAutopilotCache(user.id);
    revalidatePath("/notes");
    revalidatePath("/autopilot");
    revalidatePath("/");
    redirectWith("Note deleted.");
  }

  const updates: Partial<typeof notes.$inferInsert> = { updatedAt: new Date() };
  if (action === "favorite") updates.isFavorite = !note.isFavorite;
  else if (action === "pin") updates.isPinned = !note.isPinned;
  else if (action === "archive") updates.isArchived = true;
  else if (action === "restore") updates.isArchived = false;
  else redirectWith("Unknown note action.", "error");

  await db.update(notes).set(updates).where(eq(notes.id, note.id));
  invalidateAcademicAutopilotCache(user.id);
  revalidatePath("/notes");
  revalidatePath("/autopilot");
  revalidatePath("/");
  redirectWith("Note updated.");
}

export async function editNoteAction(formData: FormData) {
  "use server";
  try {
    await verifyCsrfToken(formData);
  } catch {
    redirectWith("Security check failed. Please try again.", "error");
  }
  const user = await currentUser();
  if (!user) redirect("/login");
  const noteId = Number(formData.get("note_id"));
  const title = sanitizePlain(formData.get("title")).slice(0, 255);
  const content = sanitizePlain(formData.get("content"));
  if (!title || !content) redirectWith("Title and content are required.", "error");
  const [note] = await db.select().from(notes).where(and(eq(notes.id, noteId), eq(notes.userId, user.id))).limit(1);
  if (!note) redirectWith("Note not found.", "error");
  const academic = await ownedAcademicSelection(user.id, formData.get("subject_id"), formData.get("topic_id"));
  const fallbackSubject = sanitizePlain(formData.get("subject")).slice(0, 128) || null;
  await db
    .update(notes)
    .set({
      title,
      content,
      subjectId: academic.subjectId,
      topicId: academic.topicId,
      subject: academic.subjectName ?? fallbackSubject,
      tags: sanitizePlain(formData.get("tags")).slice(0, 255) || null,
      isPublic: formData.get("visibility") === "public",
      updatedAt: new Date(),
    })
    .where(eq(notes.id, note.id));
  invalidateAcademicAutopilotCache(user.id);
  revalidatePath("/notes");
  revalidatePath("/autopilot");
  revalidatePath("/");
  redirectWith("Note updated.");
}

export function noteSearchPredicate(userId: number, query: string, showArchived: boolean) {
  const base = [eq(notes.userId, userId), eq(notes.isArchived, showArchived)];
  if (!query) return and(...base);
  const search = `%${query}%`;
  return and(
    ...base,
    or(ilike(notes.title, search), ilike(notes.content, search), ilike(notes.subject, search), ilike(notes.tags, search)),
  );
}

export function noteOrder() {
  return [desc(notes.isPinned), desc(notes.updatedAt), desc(notes.createdAt)] as const;
}
